cmd/config-current.go

```diff
@@ -473,12 +473,8 @@ func (s *serverConfig) loadToCachedConfigs() {
 		globalCacheExpiry = cacheConf.Expiry
 		globalCacheMaxUse = cacheConf.MaxUse
 	}
-	if globalKMS == nil {
-		globalKMSConfig = s.KMS
-		if kms, err := crypto.NewVault(globalKMSConfig); err == nil {
-			globalKMS = kms
-			globalKMSKeyID = globalKMSConfig.Vault.Key.Name
-		}
+	if err := Environment.LookupKMSConfig(s.KMS); err != nil {
+		logger.FatalIf(err, "Unable to setup the KMS")
```

Comment:
This lookup should be done from config.json only if globalKMS is not initialized.Otherwise, it will give ambiguous config error because globalKMS is already initialized.

Reply:
What is the scope of `loadToCachedConfigs()`?
From my understanding it should set the global configuration based on a cached `serverConfig` struct. Shouldn't it overwrite global config in any case, not just if it isn't set?

Reply:
When env variables for KMS have been set, globalKMS has already been initialized in loadFromEnvs. When a config entry for KMS exists, this fails with "Ambiguous KMS configuration: vault configuration and a master key are provided at the same time" because globalKMS is already initialized. So,this check needs to be done only if globalKMS is not initialized.

Reply:
When ENV vars are set, `LookupKMSConfig(...)` doesn't honor config file values. It looks up the ENV var, if present it parses and uses the value, otherwise it uses the conig file value.
`Ambiguous KMS configuration: vault configuration and a master key are provided at the same time` should only occur if the `MINIO_SSE_MASTER_KEY` **AND** an non-empty vault config is present (either through ENV or config file).

If `globalKMS != nil ` check is added the globalKMS would never be changed if config is changed and reloaded.

Reply:
@aead, this would work if LookupKMSConfig(...) is done once. To repro this issue,  start with vault env settings and no ```MINIO_SSE_MASTER_KEY``` and start minio from a fresh directory. I get a ```"Ambiguous KMS config.."``` error because the config migration creates an empty kms config in loadFromEnvs, and globalKMS gets initialized with vault config from envs. The loadToCachedConfigs is executed after that, and fails because globalKMS is not nil, and so is the server config.Here, you might want to load from cached config entries only if the globalKMS has not been initialized through env vars.

Reply:
@poornas 1. Can you retry the steps after fetching the latest version of the commit - I've fixed a bug in `LookupKMSConfig(...)`? So you set vault env and start the server. Than you kill it and restart it using a different directory (causing config creation / upgrade). With the current version I don't see this issue...

My point here is that loading the server configuration should **not** depend on the current state of the server. It should only honor all configuration options and pass if they are valid. That leads to a *stateless* LookupKMSConfig(...)` function. With `globalKMS != nil` check you skip a config lookup which may should have failed b/c of incorrect configuration.

Reply:
it works now with the latest commit @aead.